src/pretix/control/views/subevents.py

```diff
@@ -657,47 +657,59 @@ class SubEventBulkAction(SubEventQueryMixin, EventPermissionRequiredMixin, View)
+            log_entries = []
             for obj in self.get_queryset():
-                obj.log_action(
+                log_entries.append(obj.log_action(
                     'pretix.subevent.changed', user=self.request.user, data={
                         'active': False
-                    }
-                )
+                    }, save=False
+                ))
                 obj.active = False
                 obj.save(update_fields=['active'])
+
+            LogEntry.bulk_create_and_postprocess(log_entries)
             messages.success(request, pgettext_lazy('subevent', 'The selected dates have been disabled.'))
         elif request.POST.get('action') == 'enable':
+            log_entries = []
             for obj in self.get_queryset():
-                obj.log_action(
+                log_entries.append(obj.log_action(
                     'pretix.subevent.changed', user=self.request.user, data={
                         'active': True
-                    }
-                )
+                    }, save=False
+                ))
                 obj.active = True
                 obj.save(update_fields=['active'])
+
+            LogEntry.bulk_create_and_postprocess(log_entries)
             messages.success(request, pgettext_lazy('subevent', 'The selected dates have been enabled.'))
         elif request.POST.get('action') == 'delete':
             return render(request, 'pretixcontrol/subevents/delete_bulk.html', {
                 'allowed': self.get_queryset().filter(orderposition__isnull=True),
                 'forbidden': self.get_queryset().filter(orderposition__isnull=False).distinct(),
             })
         elif request.POST.get('action') == 'delete_confirm':
+            log_entries = []
+            to_delete = []
             for obj in self.get_queryset():
                 try:
                     if not obj.allow_delete():
                         raise ProtectedError('only deactivate', [obj])
-                    CartPosition.objects.filter(addon_to__subevent=obj).delete()
-                    obj.cartposition_set.all().delete()
-                    obj.log_action('pretix.subevent.deleted', user=self.request.user)
-                    obj.delete()
+                    log_entries.append(obj.log_action('pretix.subevent.deleted', user=self.request.user, save=False))
+                    to_delete.append(obj.pk)
                 except ProtectedError:
-                    obj.log_action(
+                    log_entries.append(obj.log_action(
                         'pretix.subevent.changed', user=self.request.user, data={
                             'active': False
-                        }
-                    )
+                        }, save=False,
+                    ))
                     obj.active = False
                     obj.save(update_fields=['active'])
+
+            if to_delete:
+                CartPosition.objects.filter(addon_to__subevent_id__in=to_delete).delete()
```

Comment:
This was in the original code already, but don’t addons have the subevent_id as well? I probably don’t have enough data in my local database, especially not added through the API to verify this, so I guess we better keep this?

Reply:
Weirdly, we seem to need to enforce deletion order in some cases, i.e. delete the addons first, then the regular ones, otherwise we get a foreign key issue. I think this shouldn't be necessary theoretically, but it comes up in sentry often enough, I don't think I added these for fun :\